core/ipc/sink/common/src/main/java/org/opennms/core/ipc/sink/common/AsyncDispatcherImpl.java

```diff
@@ -67,226 +65,162 @@
-                            " rejected from " +
-                            e.toString());
-                }
-            };
+            int size = asyncPolicy.getQueueSize();
+            LOG.debug("Using default in memory queue of size {}", size);
+            dispatchQueue = new DefaultQueue<>(size);
         }
 
-        state.getMetrics().register(MetricRegistry.name(state.getModule().getId(), "queue-size"), new Gauge<Integer>() {
-            @Override
-            public Integer getValue() {
-                return queue.size();
-            }
-        });
+        state.getMetrics().register(MetricRegistry.name(state.getModule().getId(), "queue-size"),
+                (Gauge<Integer>) activeDispatchers::get);
+
+        droppedCounter = state.getMetrics().counter(MetricRegistry.name(state.getModule().getId(), "dropped"));
 
-        executor = new ThreadPoolExecutor(
-                asyncPolicy.getNumThreads(),
-                asyncPolicy.getNumThreads(),
-                1000L,
-                TimeUnit.MILLISECONDS,
-                queue,
-                new LogPreservingThreadFactory(SystemInfoUtils.DEFAULT_INSTANCE_ID + ".Sink.AsyncDispatcher." + state.getModule().getId(), Integer.MAX_VALUE),
-                rejectedExecutionHandler
-            );
+        executor = Executors.newFixedThreadPool(asyncPolicy.getNumThreads(),
+                new LogPreservingThreadFactory(SystemInfoUtils.DEFAULT_INSTANCE_ID + ".Sink.AsyncDispatcher." +
+                        state.getModule().getId(), Integer.MAX_VALUE));
+        startDrainingQueue();
     }
 
-    /**
-     * When used in a ThreadPoolExecutor, this queue will block calls to
-     * {@link ThreadPoolExecutor#execute(Runnable)} when the queue is full.
-     * This is done by overriding calls to {@link LinkedBlockingQueue#offer(Object)}
-     * with calls to {@link LinkedBlockingQueue#put(Object)}, but comes with the caveat
-     * that executor must be built with <code>corePoolSize == maxPoolSize</code>.
-     * In the context of the {@link AsyncDispatcherImpl}, this is an acceptable caveat,
-     * since we enforce the matching pool sizes.
-     *
-     * There are alternative ways of solving this problem, for example we could use a
-     * {@link RejectedExecutionHandler} to achieve similar behavior, and allow
-     * for <code>corePoolSize < maxPoolSize</code>, but not for <code>corePoolSize==0</code>.
-     *
-     * For further discussions on this topic see:
-     *   http://stackoverflow.com/a/3518588
-     *   http://stackoverflow.com/a/32123535
-     *
-     * If the implementation is changed, make sure that that executor is built accordingly.
-     */
-    
-    private static class OfferBlockingQueue<E> extends LinkedBlockingQueue<E> {
-        private static final long serialVersionUID = 1L;
+    private void dispatchFromQueue() {
+        while (true) {
+            try {
+                RATE_LIMITED_LOGGER.trace("Asking dispatch queue for the next entry...");
```

Comment:
We can use the standard logger for these instead of the `RATE_LIMITED_LOGGER`. The `RATE_LIMITED_LOGGER` is intended to be used to avoid flooding the log with error messages.

Reply:
Ok should I apply this to the trace logging in `QueueFileOffHeapDispatchQueue` as well?

Reply:
Yeah, I would do the same there too.